source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -1729,380 +1734,88 @@ bool SwiftLanguageRuntime::IsValidErrorValue(
-        if (error_descriptor.m_pure_native.metadata_location == 0 ||
-            error_descriptor.m_pure_native.metadata_location ==
-                LLDB_INVALID_ADDRESS ||
-            error.Fail())
-          return false;
-        error_descriptor.m_pure_native.payload_ptr =
-            error_descriptor.m_pure_native.metadata_location;
-        error_descriptor.m_pure_native.metadata_location =
-            m_process->ReadPointerFromMemory(
-                error_descriptor.m_pure_native.payload_ptr, error);
-        if (MetadataPromiseSP promise_sp = GetMetadataPromise(
-                error_descriptor.m_pure_native.metadata_location, in_value)) {
-          if (CompilerType compiler_type = promise_sp->FulfillTypePromise()) {
-            class_type_or_name.SetCompilerType(compiler_type);
-            address.SetRawAddress(error_descriptor.m_pure_native.payload_ptr);
-            return true;
-          }
-        }
-      }
-    }
-  } break;
-  }
-
-  return false;
-}
-
 bool SwiftLanguageRuntime::GetDynamicTypeAndAddress_Protocol(
-    ValueObject &in_value, SwiftASTContext &scratch_ctx,
-    lldb::DynamicValueType use_dynamic, TypeAndOrName &class_type_or_name,
+    ValueObject &in_value, CompilerType protocol_type,
+    SwiftASTContext &scratch_ctx,
+    lldb::DynamicValueType use_dynamic,
+    TypeAndOrName &class_type_or_name,
     Address &address) {
-  CompilerType var_type(in_value.GetCompilerType());
-  SwiftASTContext::ProtocolInfo protocol_info;
-  if (!SwiftASTContext::GetProtocolTypeInfo(var_type, protocol_info))
-    return false;
+  Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
 
-  if (protocol_info.m_is_errortype)
-    return GetDynamicTypeAndAddress_ErrorType(in_value, use_dynamic,
-                                              class_type_or_name, address);
+  auto &target = m_process->GetTarget();
+  assert(IsScratchContextLocked(target) &&
+          "Swift scratch context not locked ahead");
+  auto &remote_ast = GetRemoteASTContext(scratch_ctx);
 
-  MetadataPromiseSP promise_sp;
-  ValueObjectSP instance_type_sp(
-                  in_value.GetStaticValue()->GetChildAtIndex(
-                                  protocol_info.GetInstanceTypeIndex(), true));
-  if (!instance_type_sp)
-    return false;
-  ValueObjectSP payload0_sp(
-      in_value.GetStaticValue()->GetChildAtIndex(0, true));
-  if (!payload0_sp)
-    return false;
+  lldb::addr_t existential_address;
+  bool use_local_buffer = false;
+  bool is_error = false;
 
-  auto ptr_size = m_process->GetAddressByteSize();
+  if (in_value.GetValueType() == eValueTypeConstResult) {
+    if (log)
+      log->Printf("existential value is a const result");
+
+    // Horrible hack -- we have a locally materialized value, so let
+    // our MemoryReader know where it is so that RemoteAST can read
+    // from it. Also, there's something going on with Error values
+    // I don't quite understand.
+    is_error = GetSwiftType(protocol_type)
```

Comment:
This is the precise reason I didn't switch to RemoteAST this codepath after adding the support in the compiler. Apparently the expression parser does something weird with materializing/dematerializing values, @jimingham has a rdar assigned to him about this.

Reply:
For anyone else internal the radar is rdar://38194957. While it would be nice to fix this it's not a pre-requisite since we can teach the RemoteAST memory reader to read values from known local ranges of memory too. It would clean up the code here if we didn't have to worry about that, though.